File: PracticasProgramacion/Practica_22/cpp/OrdenadorQuickSortIndirecto.cpp
#include "OrdenadorQuickSortIndirecto.h"
#include <utility>
#include <string>

int OrdenadorQuickSortIndirecto::partition(char* arr[], int low, int high) {
    char pivot = *arr[high];
    int i = low - 1;
    for (int j = low; j < high; ++j)
        if (*arr[j] <= pivot) std::swap(arr[++i], arr[j]);
    std::swap(arr[++i], arr[high]);
    return i;
}
void OrdenadorQuickSortIndirecto::quickSort(char* arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}
void OrdenadorQuickSortIndirecto::ordenarCaracteres(char* arr[], int n) { quickSort(arr, 0, n - 1); }

int OrdenadorQuickSortIndirecto::partition(int* arr[], int low, int high) {
    int pivot = *arr[high];
    int i = low - 1;
    for (int j = low; j < high; ++j)
        if (*arr[j] <= pivot) std::swap(arr[++i], arr[j]);
    std::swap(arr[++i], arr[high]);
    return i;
}
void OrdenadorQuickSortIndirecto::quickSort(int* arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}
void OrdenadorQuickSortIndirecto::ordenarEnteros(int* arr[], int n) { quickSort(arr, 0, n - 1); }

int OrdenadorQuickSortIndirecto::partition(AutoPOO* arr[], int low, int high) {
    double pivot = arr[high]->getPrecio();
    int i = low - 1;
    for (int j = low; j < high; ++j)
        if (arr[j]->getPrecio() <= pivot) std::swap(arr[++i], arr[j]);
    std::swap(arr[++i], arr[high]);
    return i;
}
void OrdenadorQuickSortIndirecto::quickSort(AutoPOO* arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}
void OrdenadorQuickSortIndirecto::ordenarAutos(AutoPOO* arr[], int n) { quickSort(arr, 0, n - 1); }

int OrdenadorQuickSortIndirecto::partition(PersonaPOO* arr[], int low, int high) {
    std::string pivot = arr[high]->getNombre();
    int i = low - 1;
    for (int j = low; j < high; ++j)
        if (arr[j]->getNombre() <= pivot) std::swap(arr[++i], arr[j]);
    std::swap(arr[++i], arr[high]);
    return i;
}
void OrdenadorQuickSortIndirecto::quickSort(PersonaPOO* arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}
void OrdenadorQuickSortIndirecto::ordenarPersonas(PersonaPOO* arr[], int n) { quickSort(arr, 0, n - 1); }

File: PracticasProgramacion/Practica_22/python/ordenador_quick_sort_indirecto.py
"""
OrdenadorQuickSortIndirecto: quick sort sobre listas de referencias.
"""

from typing import List


class AutoPOO:
    def __init__(self, precio: float = 0, anio: int = 0) -> None:
        self._precio = precio
        self._anio = anio

    def get_precio(self) -> float:
        return self._precio


class PersonaPOO:
    def __init__(self, nombre: str = "", edad: int = 0) -> None:
        self._nombre = nombre
        self._edad = edad

    def get_nombre(self) -> str:
        return self._nombre


class OrdenadorQuickSortIndirecto:
    def ordenar_caracteres(self, arr: List[str], refs: List[int]) -> None:
        if len(refs) <= 1:
            return
        pivot_val = arr[refs[-1]]
        L = [r for r in refs[:-1] if arr[r] <= pivot_val]
        R = [r for r in refs[:-1] if arr[r] > pivot_val]
        self.ordenar_caracteres(arr, L)
        self.ordenar_caracteres(arr, R)
        refs[:] = L + [refs[-1]] + R

    def ordenar_autos(self, refs: List[AutoPOO]) -> None:
        refs.sort(key=lambda x: x.get_precio())

    def ordenar_personas(self, refs: List[PersonaPOO]) -> None:
        refs.sort(key=lambda x: x.get_nombre())
